### PR TITLE
feat(campaign): add CRUD APIs

### PR DESCRIPTION
**Pull request by gary02**

### CODE REVIEW

File: src/common/utils/validator.ts [hunk 1]
```diff
@@ -129,3 +129,10 @@ export const isValidTransactionHash = (hash: string): boolean => {
 
   return /^0x([a-fA-F0-9]{64})$/.test(hash)
 }
+
+export const isValidUrl = (url: string): boolean => {
```
Comment (robertu7) [inline src/common/utils/validator.ts:133]: can align with [client-side](https://github.com/thematters/matters-web/blob/develop/src/common/utils/url.ts#L1)

File: src/connectors/campaignService.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        writingPeriod: toDatetimeRangeString(
+          writingPeriod[0],
+          writingPeriod[1]
+        ),
+        state: state || CAMPAIGN_STATE.pending,
+        creatorId,
+      },
+    })
+
+  public updateStages = async (campaignId: string, stages: Stage[]) => {
```
Comment (robertu7) [inline src/connectors/campaignService.ts:81]: use `stages` as new stages would be more simple than update the diff?

File: src/queries/campaign/index.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import description from './description'
+import name from './name'
+import participants from './participants'
+import stageName from './stage/name'
+import stages from './stages'
+
+const schema: GQLResolvers = {
+  Query: {
+    campaign: (_, { input: { shortHash } }, { dataSources: { atomService } }) =>
+      atomService.findUnique({
```
Comment (robertu7) [inline src/queries/campaign/index.ts:17]: exclude inactive campaign for non-admin users?

File: src/queries/campaign/stage/name.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import type { GQLCampaignStageResolvers } from 'definitions'
+
+const resolver: GQLCampaignStageResolvers['name'] = async (
+  { id, name },
+  _,
+  { viewer, dataSources: { translationService } }
+) => {
+  const translation = await translationService.findTranslation({
+    table: 'campaign',
```
Comment (robertu7) [inline src/queries/campaign/stage/name.ts:9]: campaign_stage

File: src/types/__test__/2/campaign.test.ts [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+    start: new Date('2024-01-01'),
+    end: new Date('2024-01-02'),
+  }
+  const writingPeriod = {
+    start: new Date('2024-01-03'),
+    end: new Date('2024-01-04'),
+  }
+  const stages = [
+    {
+      name: translations,
```
Comment (robertu7) [inline src/types/__test__/2/campaign.test.ts:86]: should use different translations as stage name 

ref: comment `src/queries/campaign/stage/name.ts#L9`

**Review by robertu7 — CHANGES_REQUESTED** [review]

File: src/connectors/campaignService.ts (hunk 2 above)
Comment (gary02) [reply src/connectors/campaignService.ts:81]: new stages will have new Ids and thus create new translation records ?

File: src/queries/campaign/stages.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import type { GQLWritingChallengeResolvers } from 'definitions'
+
+const resolver: GQLWritingChallengeResolvers['stages'] = async (
+  { id },
+  _,
+  { dataSources: { atomService } }
+) =>
+  atomService.findMany({ table: 'campaign_stage', where: { campaignId: id } })
```
Comment (robertu7) [inline src/queries/campaign/stages.ts:8]: return in order?

File: src/connectors/campaignService.ts (hunk 2 above)
Comment (robertu7) [reply src/connectors/campaignService.ts:81]: yeah, and it's more easy to [return in order](https://github.com/thematters/matters-server/pull/4026/files#r1667990887)?

**Review by robertu7 — APPROVED** [review]

**Pull request merged**